Just some tbl things.

Drup opened the pull request:
I always felt that the `Tbl` module was a ~~gross~~slight violation of the DRY principle, so this patchset removes it.
On the way, I removed almost all the duplicated `(String|Int)(Map|Set)` in the compiler, and a few others.

This is my small contribution towards making the compiler feel like one code base, instead of the collision of five different ones. ;)

## Details on `Tbl`

`Tbl` is basically a polymorphic `Map` module, using the polymorphic comparison with a few more hacks (such as `Tbl.find_str` which hardcodes the string comparison for speed. It is used for some string tables ... but not all of them). It has the exact same structure and semantics as Map, making it completely redundant.

On top of that, relying on the generic Tbl make local modifications fairly painful and invasive. For example, in the eliom patches, names are a couple of a string and a "location" (in eliom vocabulary) with a custom comparison function. With this patch, it can be implemented as a simple change in the implementation of `Env.NameMap`.

## Patchset

The first 10 commits should be fairly consensual. They were mostly done through paste and replace.

The last 3 commits, particularly 484c79b , might be less consensual: Tbl made its way into the format for cmo files though the symbol tables for global variables and primitives. I replaced them by an `Ident.Map` (for global ids) and a `StringMap` (for primitives). I didn't bump the magic number (yet).

I'm not sure if I need a change entry for this. 

## Perfs

I don't expect any perf differences (the "hot" paths were already using `Tbl.find_str`). My non-scientific benchmark show a very minor speedup (around 1%, so probably noise). Is there some command to measure performance changes of the compiler on itself?

Drup committed:
Remove Tbl in typing/

Drup committed:
Remove Tbl in bytecomp/

Drup committed:
Remove Tbl in asmcomp/

Drup committed:
Remove superfluous Sets and Maps.

Drup committed:
Remove Tbl from cmo files.

Drup committed:
Finaly remove Tbl.

Drup committed:
Depend.

Drup commented:
I do not understand travis' failure. It builds/tests fine locally and the various `depend` are up-to-date.

Octachron commented:
There is a `.depend` file hidden in ocamldoc/stdlib_non_prefixed, it seems that this the one responsible for the failure.

mshinwell commented:
I haven't got time to read this in detail right now, but will aim to do so in due course.  I agree that cleanup here would be beneficial.  This work seems related to some changes I am working on to integrate the `utils/identifiable.ml*` functionality directly into the standard library.  I will write about that within the next few weeks.

nojb commented:
>  I didn't bump the magic number (yet).

This is not necessary. Magic numbers will be bumped unconditionally for each release.

dra27 commented:
At a glance, this looks good to me (although, if I may, I think the word *gross* should possibly only be used when describing one's own work!).

It's a basically internal change, so I agree it doesn't *require* a changes entry, but that does mean the work only gets credited in Git history, so feel free to add one if you'd like!

Drup commented:
> This work seems related to some changes I am working on to integrate the utils/identifiable.ml* functionality directly into the standard library.

Indeed! Making relevant data-structures in the compiler into `Identifiable.S` would also be very nice. I almost started doing that when I realized that `Identifiable` also include printing, and printing for the source/type language is not so trivial (the printers are defined somewhere else, and require lot's of machinery).

> It's a basically internal change, so I agree it doesn't require a changes entry, but that does mean the work only gets credited in Git history, so feel free to add one if you'd like!

Ok, I will add one then.

Drup committed:
Add a change entry.

mshinwell commented on asmcomp/closure.ml:
```diff
@@ -16,12 +16,14 @@
 (* Introduction of closures, uncurrying, recognition of direct calls *)
 
 open Misc
+open Numbers
```
Instead of a whole `open`, I would just use: `module Int = Numbers.Int` here.

mshinwell commented on asmcomp/closure.ml:
```diff
@@ -16,12 +16,14 @@
 open Switch
 open Clambda
 
+module VarMap = Ident.Map
```
I think this is confusing ("ident"s versus "var"s).  How about just replacing VarMap with Ident.Map throughout?  That seems fine to me, and more clear in fact.

mshinwell commented on asmcomp/cmmgen.ml:
```diff
@@ -3397,11 +3391,7 @@ let curry_function arity =
-    type t = int
-    let compare (x:t) y = compare x y
-  end)
+module IntSet = Numbers.Int.Set
```
Here, I would use `module Int = Numbers.Int` again, then this alias can be removed.  Just use `Int.Set` instead of `IntSet`.

Drup replied:
Done.

mshinwell commented on asmcomp/selectgen.ml:
```diff
@@ -17,32 +17,33 @@
    sequentialization. *)
 
 open Misc
+open Numbers
```
Same comment as above.

mshinwell commented on bytecomp/simplif.ml:
```diff
@@ -19,6 +19,8 @@
 open Asttypes
 open Lambda
 
+module VarMap = Ident.Map
```
Same comment as above.

mshinwell commented on driver/makedepend.ml:
```diff
@@ -13,9 +13,9 @@
 (*                                                                        *)
 (**************************************************************************)
 
+open Misc
```
I suggest in `Misc` you add `Stdlib.String` and then (either via `open` or otherwise) use `String.Map` and `String.Set` instead of `StringMap` and `StringSet`; those latter two aliases in `Misc` could then be removed.  I would hope this change converges on a world where the stdlib modules gain support for sets and maps, and uniform naming can be used throughout.  (I have a patch for such stdlib support under development as I mentioned.)

Drup replied:
I just did that. That was extremely boring work. O_°

mshinwell commented on bytecomp/symtable.ml:
```diff
@@ -30,73 +32,84 @@ type error =
 exception Error of error
 
-let empty_numtable = { num_cnt = 0; num_tbl = Tbl.empty }
+module NumTbl (M : Map.S) = struct
```
General point: I think the usual style is to have a single capital letter for module names and then lowercase afterwards separated by underscores.  (This also means that if you move a module out into a file of the appropriate name, then you can keep using lowercase filenames as is conventional, yet not have to rename any of the module references in the code.)

Drup replied:
Done.

mshinwell commented on ocamltest/ocamltest_stdlib.ml:
```diff
@@ -177,8 +177,8 @@ module Sys = struct
 module StringSet = struct
-  include Set.Make (String)
+  include Misc.StringSet
   let string_of_stringset s = String.concat ", " (elements s)
```
(Following comment above.)  This should probably be [`Misc.Stdlib.`]`String.Set.to_string`.

mshinwell commented on ocamltest/ocamltest_stdlib.mli:
```diff
@@ -62,4 +62,4 @@ module StringSet : sig
 end
 
-module StringMap : Map.S with type key = string
+module StringMap = Misc.StringMap
```
Why does this alias need to be re-exported?

mshinwell replied:
(Maybe `Ocamltest_stdlib` could just be `Misc.Stdlib`?)

Drup replied:
Done.

mshinwell commented on otherlibs/unix/.depend:
```diff
@@ -77,10 +77,10 @@ cstringv.o: cstringv.c ../../byterun/caml/mlvalues.h \
  ../../byterun/caml/memory.h unixsupport.h
-dup.o: dup.c ../../byterun/caml/mlvalues.h ../../byterun/caml/config.h \
+dup2.o: dup2.c ../../byterun/caml/mlvalues.h ../../byterun/caml/config.h \
  ../../byterun/caml/m.h ../../byterun/caml/s.h ../../byterun/caml/misc.h \
```
This appears to be an unrelated change.

Drup replied:
I just made `make alldepend`. Do you want me to revert that bit?

mshinwell commented on tools/Makefile:
```diff
@@ -287,11 +287,13 @@ install::
+DUMPOBJ= \
+          ../compilerlibs/ocamlcommon.cma \
+          ../compilerlibs/ocamlbytecomp.cma \
+          \
```
Are such blank lines really necessary?

mshinwell replied:
(In fact maybe you could pull this change---the one for `tools/Makefile` that is---into a separate GPR?)

Drup replied:
I'm not sure what you mean. With this patch, dumpobj depends on `Misc`. I could try to finetune the link dependencies, but I really don't see the point for an internal tool.

mshinwell commented on typing/env.ml:
```diff
@@ -150,7 +150,7 @@ end  = struct
 end
 
-module PathMap = Map.Make(Path)
+module NameMap = Misc.StringMap
```
Names of what?

Drup replied:
`Name`, as in the individual parts in an `Longindent.t`. I would like to call them ident, but that clashes with the individual parts of a `Path.t`. I insist on having a local redefinition on that one, as it's the original motivation for this whole patch. But I agree, the name is bad.

mshinwell commented on typing/includemod.ml:
```diff
@@ -165,6 +164,12 @@ let kind_of_field_desc = function
   | Field_class _ -> "class"
   | Field_classtype _ -> "class type"
 
+(** Map indexed field names. Allow to preserve namespaces. *)
```
This comment isn't very clear---can you reword?

Drup replied:
The new one is : 
```
(** Map indexed by both field types and names. 
    This avoids name clashes between different sorts of fields 
    such as values and types. *)
```

mshinwell commented on typing/includemod.ml:
```diff
@@ -509,9 +514,9 @@ and check_modtype_equiv ~loc env ~mark cxt mty1 mty2 =
-    | Papply _ -> false
+    | Path.Pident _ -> true
+    | Path.Pdot(p, _, _) -> no_apply p
+    | Path.Papply _ -> false
```
FWIW, using disambiguation helps a lot for this kind of thing:
```
let can_alias env path =
  let rec no_apply (path : Path.t) =
    match path with
 ￼   | Pident _ -> true
    | Pdot(p, _, _) -> no_apply p
    | Papply _ -> false
  in
  ...
```

Drup replied:
Unfortunately, this part of the compiler has warning 40 turned on (and fatal). I think changing default warnings in the implementation of the typechecker shouldn't be discussed in this PR.

mshinwell commented on typing/path.ml:
```diff
@@ -107,3 +107,10 @@ let is_constructor_typath p =
+  let compare = compare
+end
+module Set = Set.Make(T)
+module Map = Map.Make(T)
```
You could probably use `Identifiable.Make` for this instead.

Drup replied:
`Identifiable.Make` require (two) printers and an equal function. I would prefer to leave that for later.

mshinwell commented on typing/typetexp.ml:
```diff
@@ -68,8 +68,10 @@ type error =
 exception Error_forward of Location.error
 
+(** Map indexed by type variable names. *)
+module VarMap = Misc.StringMap
```
Here is an example of why I recommended not having the `VarMap` alias up above in `closure.ml`: here the same alias means something completely different!  (And it's not like it's a one-letter alias.)

Drup replied:
I removed all instances of `VarMap` outside of the typechecker and renamed that one as `TyVarMap`, which should make things clearer.

mshinwell reviewed (changes requested):
I've made some comments and suggestions.  This is good work.

mshinwell commented:
Oh, and please add a `Changes` entry, in the "internal" section.  This is a fairly large diff even if mostly straightforward.

Drup committed:
Move String{Map,Set} to Misc.Stdlib.String.

Drup commented:
@mshinwell Thanks for the detailed review! I think I addressed most of your comments. I also fixed the change entry and rebased.

There are a few things (cleanup the 3 extended stdlibs, add even more identifiable everywhere) that would be very beneficial, but I would prefer to do them later, as the risk of conflict between this patchset and other changes is quite high.

Drup commented:
While replacing ocamltest's StringMap and Sets, I realized the function `stringset_of_string` was unused. So I removed it and simply use aliases to Map.Make's output.

Drup commented:
Thanks to @kit-ty-kate, the output of the opam run is [here](http://51.145.134.72:8080/). Apparently, there is an issue with annot files.

Drup committed:
Fix tests with hardcoded positions.

Drup commented:
I rebased and added some fixed related to various recent additions.

I think I already addressed all of @mshinwell 's comment. Is there a remaining blocker?

I plan to squash things before the merge.

damiendoligez commented:
You don't have to squash, just state that it should be squashed and we just use GitHub's "Squash and merge" option.

gasche commented:
"Squash and merge" is generally terrible except for PRs that obviously fit in just one commit. If it's a series of nicely separated commits ("Remove Tbl in Foo" fits the pattern) plus some junk, you have to squash manually.

On the other hand, @Drup is making a mistake by not squashing right away. He is trying to save his time, but introduces complexity, delay, communication needs and mistake points in the process instead. One might even go as far as saying as he deserves to get his entry squashed!

Drup commented:
@gasche I'm going to quote @mshinwell In another PR of mine:

> Please ensure this isn't squashed (or similar) before merging as it will be difficult to review.

New commits are not squashed immediately with the old ones so that reviewers don't have to make high order diffs between stack of patchs. Unfortunately, we're not using Iron for OCaml PR reviews.

gasche commented:
I would disagree with Mark on this one, but given that he is the one who should give you his seal of approval, it makes sense to please him.

xavierleroy commented:
"Squash and merge" is generally the right thing to do, because most PRs have an uninteresting history and uninteresting commit message, e.g. "Typos", "Going home", or "Add a Changes entry".  Indeed this PR has one commit titled "Really?!" and another titled "Depend".  I don't want to see this nor the 18 intermediate states of this PR in OCaml's commit log. I want to see one entry in the log with one message that summarizes what the PR does in its final state.

dra27 commented:
If you want to tidy the history, then I'm sure @mshinwell can be persuaded to merge a clean commit history! However, before this is merged, please could check-typo violations be dealt with:

```
./asmcomp/selectgen.ml:798.81: [long-line] line is over 80 columns
./asmcomp/selectgen.ml:1129.81: [long-line] line is over 80 columns
./bytecomp/simplif.ml:419.81: [long-line] line is over 80 columns
./bytecomp/symtable.ml:43.3: [white-at-eol] whitespace at end of line
./bytecomp/symtable.ml:62.3: [white-at-eol] whitespace at end of line
./tools/dumpobj.ml:533.81: [long-line] line is over 80 columns
./typing/includemod.ml:167.48: [white-at-eol] whitespace at end of line
./typing/includemod.ml:168.64: [white-at-eol] whitespace at end of line
./typing/includemod.ml:319.81: [long-line] line is over 80 columns
./utils/misc.ml:180.40: [white-at-eol] whitespace at end of line
./utils/misc.mli:105.7: [white-at-eol] whitespace at end of line
```

damiendoligez commented:
It is generally a bad idea to squash before the end because the commits get out of sync with the discussion and it becomes hard to make sense of them.

Drup commented:
@dra27 All fixed, and rebased on top of trunk to take advantage of the nice typocleaning.

dra27 commented:
Thanks for the `check-typo` loveliness 🙂

Drup commented:
4.07 is released, the world cup is finished. @mshinwell, can I get a review? :)

mshinwell commented:
Please add a comment in env.ml on the "NameMap" definition to say what "Name" means here, since we can't come up with anything better.  OK to merge with that change.  Thanks for taking the time to do this, it seems like a real improvement.

Drup commented:
Done, rebased, and history somewhat cleaned up.

Drup commented:
CI is green.

gasche commented:
I'm a bit worried by the last commit fixing a test failure. This means that the testsuite fails if I am bisecting and end in the middle of your patchset, right? Could you maybe fixup this commit at the first point where the change is required, to maintain the invariant that intermediate states are clean?

mshinwell commented:
I was planning to squash the whole lot; I'm not sure any intermediate state is useful here.

Drup commented:
@mshinwell Do as you wish. :p
